WalletWasabi.Fluent.Generators/UIContextGenerator.cs

```diff
@@ -0,0 +1,326 @@
+	{
+		var result = new List<ConstructorDeclarationSyntax>();
+
+		var (compilation, items) = args;
+
+		var toGenerate =
+			from i in items
+			let cls = i.Node as ClassDeclarationSyntax
+			where cls != null
+			group i by cls.Identifier.ValueText into g
+			select g.First();
+
+		foreach (var item in toGenerate)
+		{
+			var (node, model) = (item.Node, item.SemanticModel);
+			if (node is not ClassDeclarationSyntax classDeclaration)
+			{
+				continue;
+			}
+
+			if (model.GetDeclaredSymbol(classDeclaration) is not INamedTypeSymbol classSymbol)
+			{
+				continue;
+			}
+
+			var ctors = GenerateConstructors(context, classDeclaration, model, classSymbol).ToList();
+
+			result.AddRange(ctors);
+		}
+
+		return result;
+	}
+
+	private static IEnumerable<ConstructorDeclarationSyntax> GenerateConstructors(SourceProductionContext context, ClassDeclarationSyntax classDeclaration, SemanticModel semanticModel, INamedTypeSymbol classSymbol)
+	{
+		var fileName = classDeclaration.Identifier.ValueText + UiContextAnalyzer.UiContextFileSuffix;
+
+		var className = classSymbol.Name;
+		var namespaceName = classSymbol.ContainingNamespace.ToDisplayString();
+
+		var ctors =
+			classDeclaration.ChildNodes()
+							.OfType<ConstructorDeclarationSyntax>()
+							.ToList();
+
+		foreach (var ctor in ctors)
+		{
+			if (!classDeclaration.GetUiContextReferences(semanticModel).Any())
+			{
+				if (!classDeclaration.IsAbstractClass(semanticModel) && ctor.IsPublic())
+				{
+					yield return ctor;
+				}
+			}
+			// if constructor already has a UIContext parameter, leave it be. Don't generate a new ctor and use the current one for FluentNavigation.
+			else if (ctor.ParameterList.Parameters.Any(p => p.Type.IsUiContextType(semanticModel)))
+			{
+				// it must be public though
+				if (ctor.IsPublic())
+				{
+					yield return ctor;
+				}
+			}
+			else
+			{
+				var ctorArgs =
+					  ctor.ParameterList.Parameters
+						  .Select(x => x.Identifier.ValueText)
+						  .ToList();
+
+				var hasCtorArgs = ctorArgs.Any();
+				var ctorArgsString = string.Join(",", ctorArgs);
+				var ctorString =
+					hasCtorArgs
+					? $": this({ctorArgsString})"
+					: "";
+
+				var parameterUsings =
+					ctor.ParameterList.Parameters
+									  .Where(p => p.Type is not null)
+									  .Select(p => semanticModel.GetTypeInfo(p.Type!))
+									  .Where(t => t.Type is not null)
+									  .Select(t => $"using {t.Type!.ContainingNamespace.ToDisplayString()};")
+									  .ToList();
+
+				var parametersString = "UiContext uiContext";
+
+				var uiContextParameter =
+					SyntaxFactory.Parameter(SyntaxFactory.Identifier("uiContext")
+														 .WithLeadingTrivia(SyntaxFactory.Space))
+								 .WithType(SyntaxFactory.ParseTypeName("UiContext"));
+
+				if (hasCtorArgs)
+				{
+					parametersString += ", ";
+				}
+
+				parametersString =
+					ctor.ParameterList.Parameters.Insert(0, uiContextParameter).ToFullString();
```

Comment:
@ichthus1604 
I just realized while reviewing #12 that something smells here.

```
if (hasCtorArgs)
				{
					parametersString += ", ";
				}
```

Doesn't make sense as the next line overwrites the value.

Can you take a look?